Fixes #145: improve the code display (double and triple backticks)

pacamara opened the pull request:
Fixes #145: "improve the code display (double and triple backticks)"

Introduces a new ROW_TYPE_CODE for messages that contain a mixture of fenced (i.e. 3 backtick) code blocks and text sections. The text sections can contain in-line (unfenced) code blocks. Each of the fenced code blocks is in its own HorizontalScrollView and independently scrollable.

This is done via a new layout adapter_item_vector_message_code.xml, which can contain a mixture of  adapter_item_vector_message_code_block.xml and adapter_item_vector_message_code_text.xml layouts. 

A different background colour for fenced code blocks can be specified using the new attribute fenced_code_block_background_color. The current colour is by way of example only, and is intended to be changed by designers.

![issue145_fixed_screenie_postreview](https://user-images.githubusercontent.com/19371062/34157064-da8ead72-e4b7-11e7-941f-33ef1208be8c.png)

Screenshot showing a single message containing:
* two fenced (3 backtick) code blocks, one of which has been scrolled
* a plain text section containing inline code
* plain text sections with no code

ylecollen commented on vector/src/main/java/im/vector/adapters/VectorMessagesAdapterHelper.java:
````diff
@@ -552,6 +552,45 @@ public void onClick(View view) {
+     * handling (issue 145) */
+    public int getTickCount(final Message message) {
+        final String mb = message.body;
+        if (mb.startsWith("```") && mb.endsWith("```")) {
````
1 - what is the plan is the message contains several blocks of code ?
you only manage 1 block code by message.

2 - the markdown might be disabled.

pacamara replied:
1. Ah, I didn't know this was a requirement. In that case, if the message contains one or more code blocks, use the adapter_item_vector_message_code.xml  layout, and use spans to set the colour of each block.
2. Will fix this.

pacamara replied:
Issues 1 and 2 both fixed now.

ylecollen commented on vector/src/main/java/im/vector/adapters/VectorMessagesAdapterHelper.java:
```diff
@@ -552,6 +552,45 @@ public void onClick(View view) {
+            return 1;
+        }
+        else {
+            return 0;
```
give a name for ``, ``, ` block

pacamara replied:
In my implementation of your review, this function no longer exists.

ylecollen commented on vector/src/main/res/layout/adapter_item_vector_message_code.xml:
```diff
@@ -0,0 +1,121 @@
+<?xml version="1.0" encoding="utf-8"?>
```
i suspect it is the same layout than the text message

pacamara replied:
It's different from adapter_item_vector_search_message_text_emote_notice.xml in three ways:  
i) wraps the TextView in a HorizontalScrollView
ii) monospace font
iii) no auto-linking of URLs

JonasDralle commented:
This fixes the problem of newlines being ignored, right?

pacamara commented:
@JonasDralle Yes, eg in the HelloWorld code example in the screenshot, code is not wrapped and newlines are not ignored.  
@ylecollen Thanks for the review!

pacamara commented:
@ylecollen @JonasDralle  I've implemented @ylecollen 's code review and pushed it as a separate commit. Will squash to a single commit when it's time to merge. A single message can now contain multiple fenced (3 backtick) code blocks each of which lives in its own independently scrollable view, as well as text sections which can contain inline code blocks. It also respects the markdown-enabled config setting now.

(I've removed support for a separate colour for inline code blocks since this is already controlled by an existing Riot attribute. Please let me know if you sitll want support for a separate colour).

ylecollen commented:
LGTM

ylecollen commented:
@pacamara 
could you provide us more info we add you in the contributors list ?

https://github.com/vector-im/riot-android/blob/master/AUTHORS.rst

pacamara commented:
@ylecollen Thanks! :)  Contributor info is:
Joseph Gittings <joe.spbn at gmail dot com>

It says in Status Open Bounty, "payout pending maintainer confirmation", so my payout is still pending, right? I have updated my SNT receiving address in SOB to the correct value.

ylecollen commented:
@pacamara 
it does not fill all the cases : for example, the https://imgur.com/3Hb4hgh is not fixed.
ie if you send a block of code from the webclient, the formatted body could contain the <pre><code>.. tags.

ylecollen commented:
improve with
https://github.com/vector-im/riot-android/commits/develop

but it does not fix all the cases.

pacamara commented:
@ylecollen Hi! Was chatting about this with @lampholder yesterday and we came up with these two to-do items to address the above concerns:

a) apply markdown-enabled setting at message send time, not display time.
b)  look for <code> and <pre><code> blocks in message. formatted_body rather than backticks in message. body, and split the blocks[] array on those.

Point (b) seems a simple enough change and would address compatibility between the web and android clients. I'm happy to do both changes or would you rather?

ylecollen commented:
@pacamara 
i only patched your code until you have time to fix everything properly.
i've got some comments that the code block height and its background colour.

pacamara commented:
@ylecollen Cool, I'll get on this and build on your changes then.  :+1:

pacamara commented:
@ylecollen  OK done! See https://github.com/vector-im/riot-android/pull/1878  :beers:
